### PR TITLE
Initial implementations for BPF-based socket and process events tables

### PR DESCRIPTION
**Pull request by alessandrogario**
This PR implements two new tables: **bpf_process_events**, **bpf_socket_events** using the ebpfpub library from https://github.com/trailofbits/ebpfpub.

**BPF support requires a kernel >= 4.18!**

VMware Fusion users: due to [how VMware works](https://github.com/trailofbits/ebpfpub/issues/42), you will either have to
1. Lower memory usage: pass **bpf_buffer_storage_size** and **bpf_perf_event_array_exp**
2. Set **vcpu.hotadd** to `FALSE` in your vmx file (**PREFERRED**) - thanks @mike-myers-tob !

# TLDR
 * Plain VMware Fusion: `sudo ./osqueryi --verbose --disable_events=false --enable_bpf_events=true --bpf_buffer_storage_size=512 --bpf_perf_event_array_exp=10`
 * Anything else: `sudo ./osqueryi --verbose --disable_events=false --enable_bpf_events=true`

Replaces https://github.com/osquery/osquery/pull/6337

### CODE REVIEW

**Comment by theopolis** [comment]
Can you separate out the bug fixes into another PR so we can land those quickly?

**Comment by alessandrogario** [comment]
> Can you separate out the bug fixes into another PR so we can land those quickly?

The include order commit is necessary by BPF to compile correctly, otherwise the build will break locally. I am not sure why this error is hidden in the CI, since those C headers break all the other headers by defining several macros

File: osquery/main/main.cpp [hunk 1]
```diff
@@ -178,6 +182,13 @@ int startOsquery(int argc, char* argv[]) {
 
   // Only worker processes should start a daemon or shell.
   if (!runner.isWatcher()) {
+#if defined(__unix__) && !defined(__APPLE__)
+    auto status = configureBPFMemoryLimits();
```
Comment (theopolis) [inline osquery/main/main.cpp:186]: Could this instead be called within the BPF publishers? For example each publisher could call this configure API and within this method you could implement a `std::call_once` to set the RLIMIT.

**Comment by mike-myers-tob** [comment]
As a user I would prefer to just have the one `process_events` table that automatically detects and uses the best-available event source.

If I may boldly suggest, could the `process_events` table's underlying implementation be defaulted to eBPF and then these config switches could be an override for that?

**Comment by alessandrogario** [comment]
> As a user I would prefer to just have the one `process_events` table that automatically detects and uses the best-available event source.
> 
> If I may boldly suggest, could the `process_events` table's underlying implementation be defaulted to eBPF and then these config switches could be an override for that?

Given how long it has taken to get this feature to the current state, I'd rather not add additional feature requests to it and merge it as soon as possible.

There are many reasons (IMHO) to keep it separate:
1. Audit, BPF, BSM, EndpointSecurity all have different event fields they can capture, with different data integrity guarantees. Even though they may look the same, I wouldn't mix them inside the same table right now
2. Switching the same table between multiple publisher implementations has never been done, and we should first make this available as a feature in osquery core and provide tests for it
3. I think it is useful to be able to use both publishers at the same time while we test the functionality and look for feedback.  This allows user to opt in to the BPF testing without necessarily deciding to stop collecting data with a more stable method

**Comment by mike-myers-tob** [comment]
Ok I retract my suggestion!

**Comment by theopolis** [comment]
I am happy to code review when you are ready to consider merging.

**Comment by alessandrogario** [comment]
I think I am happy with what I have in the branch, looking for code reviews!

File: osquery/events/linux/bpf/uniquedir.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/uniquedir.h:2]: TODO: Update copyright header

File: osquery/events/tests/linux/bpf/mockedfilesystem.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/tests/linux/bpf/mockedfilesystem.cpp:2]: TODO: Update copyright header

File: osquery/events/tests/linux/bpf/mockedfilesystem.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/tests/linux/bpf/mockedfilesystem.h:2]: TODO: Update copyright header

File: osquery/events/tests/linux/bpf/mockedprocesscontextfactory.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/tests/linux/bpf/mockedprocesscontextfactory.cpp:2]: TODO: Update copyright header

File: osquery/events/tests/linux/bpf/mockedprocesscontextfactory.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/tests/linux/bpf/mockedprocesscontextfactory.h:2]: TODO: Update copyright header

File: osquery/tables/events/linux/bpf_process_events.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/tables/events/linux/bpf_process_events.cpp:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/filesystem.cpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/filesystem.cpp:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/filesystem.h [hunk 9]
```diff
@@ -0,0 +1,42 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/filesystem.h:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/ifilesystem.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/ifilesystem.h:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/iprocesscontextfactory.h [hunk 11]
```diff
@@ -0,0 +1,69 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/iprocesscontextfactory.h:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/isystemstatetracker.h [hunk 12]
```diff
@@ -0,0 +1,128 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/isystemstatetracker.h:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/processcontextfactory.cpp [hunk 13]
```diff
@@ -0,0 +1,278 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/processcontextfactory.cpp:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/processcontextfactory.h [hunk 14]
```diff
@@ -0,0 +1,47 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/processcontextfactory.h:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/setrlimit.cpp [hunk 15]
```diff
@@ -0,0 +1,49 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/setrlimit.cpp:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/setrlimit.h [hunk 16]
```diff
@@ -0,0 +1,13 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/setrlimit.h:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/systemstatetracker.cpp [hunk 17]
```diff
@@ -0,0 +1,1016 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/systemstatetracker.cpp:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/systemstatetracker.h [hunk 18]
```diff
@@ -0,0 +1,218 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/systemstatetracker.h:2]: TODO: Update copyright header

File: osquery/tables/events/linux/bpf_process_events.h [hunk 19]
```diff
@@ -0,0 +1,33 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/tables/events/linux/bpf_process_events.h:2]: TODO: Update copyright header

File: osquery/tables/events/linux/bpf_socket_events.cpp [hunk 20]
```diff
@@ -0,0 +1,109 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/tables/events/linux/bpf_socket_events.cpp:2]: TODO: Update copyright header

File: osquery/tables/events/linux/bpf_socket_events.h [hunk 21]
```diff
@@ -0,0 +1,27 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/tables/events/linux/bpf_socket_events.h:2]: TODO: Update copyright header

File: libraries/cmake/source/ebpfpub/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline libraries/cmake/source/ebpfpub/CMakeLists.txt:1]: TODO: Update copyright header

File: libraries/cmake/source/modules/Findebpfpub.cmake [hunk 23]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline libraries/cmake/source/modules/Findebpfpub.cmake:1]: TODO: Update copyright header

File: osquery/events/linux/bpf/bpfeventpublisher.cpp [hunk 24]
```diff
@@ -0,0 +1,1019 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/bpfeventpublisher.cpp:2]: TODO: Update copyright header

File: osquery/events/linux/bpf/bpfeventpublisher.h [hunk 25]
```diff
@@ -0,0 +1,169 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (alessandrogario) [inline osquery/events/linux/bpf/bpfeventpublisher.h:2]: TODO: Update copyright header

**Comment by mike-myers-tob** [comment]
> I am happy to code review when you are ready to consider merging.

Ok Alessandro seems to be ready for your review help, as you are available. (Thanks!)

File: libraries/cmake/source/thrift/CMakeLists.txt [hunk 26]
```diff
@@ -87,6 +87,16 @@ function(thriftMain)
     )
   endif()
 
+  if(PLATFORM_WINDOWS)
+    set(forced_include_file_flag "/FI")
+  else()
+    set(forced_include_file_flag "-include")
+  endif()
+
+  target_compile_options(thirdparty_thrift PRIVATE
```
Comment (theopolis) [inline libraries/cmake/source/thrift/CMakeLists.txt:96]: What is the story here? And can you leave a note in code about this?

File: specs/linux/bpf_process_events.table [hunk 27]
```diff
@@ -0,0 +1,22 @@
+table_name("bpf_process_events")
+description("Track time/action process executions.")
+schema([
+    Column("bpf_time", TEXT, "The nsecs uptime timestamp as obtained from BPF"),
```
Comment (theopolis) [inline specs/linux/bpf_process_events.table:4]: Thoughts on naming this `ntime` since it is `time` but in `nsecs`?

File: specs/linux/bpf_process_events.table [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+    Column("thread_id", BIGINT, "Thread ID"),
```
Comment (theopolis) [inline specs/linux/bpf_process_events.table:5]: I recommend using `tid` since we've set that precedent.

File: specs/linux/bpf_process_events.table [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+    Column("process_id", BIGINT, "Process ID"),
```
Comment (theopolis) [inline specs/linux/bpf_process_events.table:6]: I recommend using `pid` since we've set that precedent.

File: specs/linux/bpf_process_events.table [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+    Column("parent_process_id", BIGINT, "Parent process ID"),
```
Comment (theopolis) [inline specs/linux/bpf_process_events.table:7]: I recommend using `parent` since we've set that precedent.

File: specs/linux/bpf_process_events.table [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+    Column("uid", BIGINT, "User ID"),
+    Column("gid", BIGINT, "Group ID"),
+    Column("cgroup_id", INTEGER, "Cgroup ID"),
```
Comment (theopolis) [inline specs/linux/bpf_process_events.table:10]: Thoughts on using `cid`?

File: specs/linux/bpf_process_events.table [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+    Column("exit_code", TEXT, "Exit code of the system call"),
+    Column("probe_error", INTEGER, "Set to 1 if one or more buffers could not be captured"),
+    Column("syscall", TEXT, "System call name"),
+    Column("binary_path", TEXT, "Binary path"),
```
Comment (theopolis) [inline specs/linux/bpf_process_events.table:14]: I recommend using `path` since we've set that precedent.

File: osquery/events/tests/linux/bpf/utils.cpp [hunk 33]
```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include "utils.h"
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply osquery/events/tests/linux/bpf/utils.cpp:10]: The headers inside the tests are not exported and are not accessible using the relative path, and I don't think include namespaces have been used with tests yet

File: osquery/events/linux/bpf/filesystem.cpp [hunk 34]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  if (buffer.empty()) {
+    return false;
+  }
+
+  return true;
+}
+
+bool Filesystem::enumFiles(int dirfd, EnumFilesCallback callback) const {
```
Comment (theopolis) [inline osquery/events/linux/bpf/filesystem.cpp:98]: Check if `std::filesystem` has an implementation of this directory walk that you can reuse.

File: osquery/events/linux/bpf/processcontextfactory.cpp [hunk 35]
```diff
@@ -0,0 +1,279 @@
...
+  auto succeeded = fs.enumFiles(
+    process_fdmap.get(),
+
+    [&](const std::string &name, bool directory) {
+      if (directory) {
+        return;
+      }
+
+      char* null_terminator{nullptr};
+      auto int_fd_value = std::strtoull(name.c_str(), &null_terminator, 10);
```
Comment (theopolis) [inline osquery/events/linux/bpf/processcontextfactory.cpp:81]: Can you see if you can perform this conversion with one of our `tryTo<unsigned long long>` utilities?

File: osquery/events/linux/bpf/processcontextfactory.cpp [hunk 36]
```diff
@@ -0,0 +1,279 @@
...
+  auto succeeded = fs.enumFiles(
+    process_root.get(),
+
+    [&](const std::string &name, bool directory) {
+      if (!directory) {
+        return;
+      }
+
+      char* null_terminator{nullptr};
+      auto pid = static_cast<pid_t>(std::strtoull(name.c_str(), &null_terminator, 10));
```
Comment (theopolis) [inline osquery/events/linux/bpf/processcontextfactory.cpp:174]: Ditto with `tryTo`.

File: osquery/events/linux/bpf/processcontextfactory.cpp [hunk 37]
```diff
@@ -0,0 +1,279 @@
...
+  auto parent_pid_it = std::next(parent_pid_r_it.base(), 4);
+  if (parent_pid_it >= buffer.end()) {
+    return false;
+  }
+
+  auto parent_pid_index = parent_pid_it - buffer.begin();
+  auto parent_pid_ptr = &buffer[parent_pid_index];
+
+  char* space_separator_ptr{nullptr};
+  auto output_pid = std::strtoull(parent_pid_ptr, &space_separator_ptr, 10);
```
Comment (theopolis) [inline osquery/events/linux/bpf/processcontextfactory.cpp:251]: Ditto `tryTo`.

File: osquery/events/linux/bpf/systemstatetracker.cpp [hunk 38]
```diff
@@ -0,0 +1,1017 @@
...
+}
+
+SystemStateTracker::Ref SystemStateTracker::create(
+    IProcessContextFactory::Ref process_context_factory) {
+  try {
+    return SystemStateTracker::Ref(
+        new SystemStateTracker(std::move(process_context_factory)));
+
+  } catch (const Status& status) {
+    LOG(ERROR) << status.getMessage();
```
Comment (theopolis) [inline osquery/events/linux/bpf/systemstatetracker.cpp:46]: I generally do not like this pattern, I think we this should add a tiny bit more context:

```
LOG(ERROR) << "Could not create state tracker: " << status.getMessage();
```

File: osquery/events/linux/bpf/systemstatetracker.cpp [hunk 39]
```diff
@@ -0,0 +1,1017 @@
...
+    const std::string& path) {
+  auto& process_context =
+      getProcessContext(context, process_context_factory, process_id);
+
+  if (path.front() == '/') {
+    process_context.cwd = path;
+
+  } else {
+    if (process_context.cwd.back() != '/') {
+      process_context.cwd += "/";
```
Comment (theopolis) [inline osquery/events/linux/bpf/systemstatetracker.cpp:391]: can you `+= '/';`?

File: osquery/events/linux/bpf/systemstatetracker.cpp [hunk 40]
```diff
@@ -0,0 +1,1017 @@
...
+      getProcessContext(context, process_context_factory, process_id);
+
+  std::string absolute_path;
+  if (path.front() == '/') {
+    absolute_path = path;
+
+  } else if (dirfd == AT_FDCWD) {
+    absolute_path = process_context.cwd;
+    if (absolute_path.back() != '/') {
+      absolute_path += "/";
```
Comment (theopolis) [inline osquery/events/linux/bpf/systemstatetracker.cpp:422]: can you `+= '/';`?

File: osquery/events/linux/bpf/systemstatetracker.cpp [hunk 41]
```diff
@@ -0,0 +1,1017 @@
...
+            fd_info.data)) {
+      return false;
+    }
+
+    const auto& file_data =
+        std::get<ProcessContext::FileDescriptor::FileData>(fd_info.data);
+
+    absolute_path = file_data.path;
+    if (absolute_path.back() != '/') {
+      absolute_path += "/";
```
Comment (theopolis) [inline osquery/events/linux/bpf/systemstatetracker.cpp:444]: can you `+= '/';`?

File: osquery/events/linux/bpf/systemstatetracker.cpp [hunk 42]
```diff
@@ -0,0 +1,1017 @@
...
+
+  sockaddr_in addr{};
+  std::memcpy(&addr, sockaddr.data(), size);
+
+  port = static_cast<std::uint16_t>(ntohs(addr.sin_port));
+
+  std::uint8_t components[4];
+  std::memcpy(components, &addr.sin_addr.s_addr, sizeof(components));
+
+  address = std::to_string(components[0]) + ".";
```
Comment (theopolis) [inline osquery/events/linux/bpf/systemstatetracker.cpp:890]: can you `+ '.';`?

Comment (theopolis) [reply osquery/events/linux/bpf/systemstatetracker.cpp:890]: and ditto for the next two lines?

**Comment by theopolis** [comment]
This is some great code!

One thing I noticed throughout that is a nitpick, I see in a lot of locations you are applying the style

```
namespace osquery {
CODE HERE
}
```

And I think we should be consistent and use:

```
namespace osquery {

CODE HERE

}
```

Where we have a blank newline between each.

Heads up that we should give folks a heads up when this is merged that they need to update their toolchain. It is very possible they are still using the original toolchain since nothing has explicitly required them to update. I think the best medium for this is to include a notice in the commit body and then we can mention+pin in Slack.

With this I have completed my code review. I will continue some run time testing on different setups I can find and report back any errors or questions.

**Review by theopolis — CHANGES_REQUESTED** [review]

**Comment by theopolis** [comment]
Should I expect duplicated events when running `sudo su`?

```
osquery> select * from bpf_process_events where pid = 25664;
+---------------+-------+-------+--------+------+------+------+----------------------+-------------+---------+-----------------------------+-------------------------------+---------+----------+
| bpf_ntime     | tid   | pid   | parent | uid  | gid  | cid  | exit_code            | probe_error | syscall | path                        | cwd                           | cmdline | duration |
+---------------+-------+-------+--------+------+------+------+----------------------+-------------+---------+-----------------------------+-------------------------------+---------+----------+
| 4942316608852 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /home/teddy/.cargo/bin/sudo | /home/teddy/git/osquery/build | sudo su | 10320    |
| 4942316629679 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /home/teddy/gocode/bin/sudo | /home/teddy/git/osquery/build | sudo su | 4331     |
| 4942316641199 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/local/go/bin/sudo      | /home/teddy/git/osquery/build | sudo su | 3773     |
| 4942316651335 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/local/sbin/sudo        | /home/teddy/git/osquery/build | sudo su | 3279     |
| 4942316662844 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/local/bin/sudo         | /home/teddy/git/osquery/build | sudo su | 3663     |
| 4942316678115 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/sbin/sudo              | /home/teddy/git/osquery/build | sudo su | 4414     |
| 4942316687115 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 0                    | 0           | exec    | /usr/bin/sudo               | /home/teddy/git/osquery/build | sudo su | 176637   |
+---------------+-------+-------+--------+------+------+------+----------------------+-------------+---------+-----------------------------+-------------------------------+---------+----------+
```

My expectation when using the table was to see one event generated.

**Comment by alessandrogario** [comment]
> Should I expect duplicated events when running `sudo su`?
> 
> ```
> osquery> select * from bpf_process_events where pid = 25664;
> +---------------+-------+-------+--------+------+------+------+----------------------+-------------+---------+-----------------------------+-------------------------------+---------+----------+
> | bpf_ntime     | tid   | pid   | parent | uid  | gid  | cid  | exit_code            | probe_error | syscall | path                        | cwd                           | cmdline | duration |
> +---------------+-------+-------+--------+------+------+------+----------------------+-------------+---------+-----------------------------+-------------------------------+---------+----------+
> | 4942316608852 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /home/teddy/.cargo/bin/sudo | /home/teddy/git/osquery/build | sudo su | 10320    |
> | 4942316629679 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /home/teddy/gocode/bin/sudo | /home/teddy/git/osquery/build | sudo su | 4331     |
> | 4942316641199 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/local/go/bin/sudo      | /home/teddy/git/osquery/build | sudo su | 3773     |
> | 4942316651335 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/local/sbin/sudo        | /home/teddy/git/osquery/build | sudo su | 3279     |
> | 4942316662844 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/local/bin/sudo         | /home/teddy/git/osquery/build | sudo su | 3663     |
> | 4942316678115 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 18446744073709551614 | 0           | exec    | /usr/sbin/sudo              | /home/teddy/git/osquery/build | sudo su | 4414     |
> | 4942316687115 | 25664 | 25664 | 25569  | 1000 | 1000 | 2311 | 0                    | 0           | exec    | /usr/bin/sudo               | /home/teddy/git/osquery/build | sudo su | 176637   |
> +---------------+-------+-------+--------+------+------+------+----------------------+-------------+---------+-----------------------------+-------------------------------+---------+----------+
> ```
> 
> My expectation when using the table was to see one event generated.

The table is currently showing failed executions, which are happening because the shell is attempt to do something like
```
for (const auto &path : ENV["PATH"]) {
  exec(path + '/' + sudo)
}
```

I'll remove them from the table, and maybe we can add an option to enabled them back on if a user wishes to see those rows

File: cmake/flags.cmake [hunk 43]
```diff
@@ -19,7 +19,7 @@ function(setupBuildFlags)
   add_library(cxx_settings INTERFACE)
   add_library(c_settings INTERFACE)
 
-  target_compile_features(cxx_settings INTERFACE cxx_std_14)
+  target_compile_features(cxx_settings INTERFACE cxx_std_17)
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply cmake/flags.cmake:22]: There are many C++17 things used in the BPF libraries such as std::optional, std::holds_alternative, ...

File: osquery/utils/tests/rot13.cpp [hunk 44]
```diff
@@ -1,9 +1,10 @@
 /**
- *  Copyright (c) 2014-present, Facebook, Inc.
- *  All rights reserved.
+ * Copyright (c) 2014-present, The osquery authors
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply osquery/utils/tests/rot13.cpp:2]: Oops! I thought it was a leftover from my old base when I corrected all the headers! Will revert this change in my branch

File: osquery/tables/system/linux/deb_packages.cpp [hunk 45]
```diff
@@ -26,6 +19,13 @@ extern "C" {
 #include <osquery/worker/ipc/platform_table_container_ipc.h>
 #include <osquery/worker/logging/glog/glog_logger.h>
 
+extern "C" {
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply osquery/tables/system/linux/deb_packages.cpp:22]: As soon as I added the BPF libraries and switched to C++17, these C headers started preventing the build from passing due to several definitions causing errors in the other C++ headers

Several things changed since then; I'll see if it can be removed

**Comment by theopolis** [comment]
Also it looks like the exit code is a signed value and the table is reporting as unsigned (just a guess from observation).

File: osquery/events/linux/bpf/bpfeventpublisher.cpp [hunk 46]
```diff
@@ -0,0 +1,990 @@
...
+
+  while (!isEnding()) {
+    d->perf_event_reader->exec(
+        std::chrono::seconds(1U),
+
+        [&](const tob::ebpfpub::IFunctionTracer::EventList& event_list,
+            const tob::ebpfpub::IPerfEventReader::ErrorCounters&
+                error_counters) {
+          if (error_counters.invalid_probe_output != 0U) {
+            VLOG(1) << "invalid_probe_output: "
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply osquery/events/linux/bpf/bpfeventpublisher.cpp:203]: The VLOG(1) already prefixes the file name to the log message, which always match the class name; should we still add the class name at the start of the message?

The newline character is something I forgot to remove when converting this from `std::cerr` to `VLOG(1)`! Will remove it right away

File: osquery/events/linux/bpf/filesystem.cpp (hunk 34 above)
Comment (alessandrogario) [reply osquery/events/linux/bpf/filesystem.cpp:98]: I am using fd-based functions to reduce race conditions when traversing paths; the std::filesystem implementation would not support this

File: osquery/events/linux/bpf/processcontextfactory.cpp (hunk 37 above)
Comment (alessandrogario) [reply osquery/events/linux/bpf/processcontextfactory.cpp:251]: The std::strtoull in this case is required because we are relying on finding the space separator; tryTo does not support this and will return an error

File: osquery/events/linux/bpf/systemstatetracker.cpp (hunk 38 above)
Comment (alessandrogario) [reply osquery/events/linux/bpf/systemstatetracker.cpp:46]: > Also it looks like the exit code is a signed value and the table is reporting as unsigned (just a guess from observation).

Yes, the library behind always work with std::uint64_t values as it's the most generic type (which also works well when pointers are returned by the traced functions).

File: osquery/events/linux/bpf/bpfeventpublisher.cpp (hunk 46 above)
[the remark this answers is not in the document]
Comment (theopolis) [reply osquery/events/linux/bpf/bpfeventpublisher.cpp:206]: I don't think we should add the class name per-say but we should add some context and not rely on the stdout formatting of the file/line number as the logger plugins separate these as distinct fields. If you are viewing these verbose messages in splunk you want the `message` row to have meaningful context.

File: osquery/tables/events/linux/bpf_process_events.cpp [hunk 47]
```diff
@@ -42,6 +42,11 @@ bool BPFProcessEventSubscriber::generateRow(
     return false;
   }
 
+  auto signed_exit_code = static_cast<std::int64_t>(event.bpf_header.exit_code);
+  if (signed_exit_code == -1LL) {
```
Comment (theopolis) [inline osquery/tables/events/linux/bpf_process_events.cpp:46]: I think you want to check for -2:

```python
>>> import ctypes
>>> k = -2
>>> ctypes.c_ulong(k).value
18446744073709551614L
```

**Review by theopolis — CHANGES_REQUESTED** [review]

File: osquery/tables/system/linux/deb_packages.cpp (hunk 45 above)
[the remark this answers is not in the document]
Comment (alessandrogario) [reply osquery/tables/system/linux/deb_packages.cpp:22]: I don't think it's possible to remove this and still be able to compile things; there are things inside those headers that will just break the build such as this macro

```
../libraries/cmake/source/libdpkg/src/lib/dpkg/macros.h:127:9: note: macro 'clamp' defined here
#define clamp(v, l, h) ((v) > (h) ? (h) : ((v) < (l) ? (l) : (v)))
```

colliding with [std::clamp](https://en.cppreference.com/w/cpp/algorithm/clamp)

File: osquery/tables/events/tests/linux/bpf_socket_events_tests.cpp [hunk 48]
```diff
@@ -0,0 +1,130 @@
...
+
+#include <gtest/gtest.h>
+
+#include <osquery/tables/events/linux/bpf_socket_events.h>
+
+#include <netinet/in.h>
+
+namespace osquery {
+namespace {
+// clang-format off
```
[the remark this answers is not in the document]
Comment (alessandrogario) [reply osquery/tables/events/tests/linux/bpf_socket_events_tests.cpp:20]: It doesn't look as pretty but it's still readable, so I'll remove it!

File: osquery/tables/events/linux/bpf_process_events.cpp (hunk 47 above)
Comment (alessandrogario) [reply osquery/tables/events/linux/bpf_process_events.cpp:46]: I mistakenly used the exit code documented in the glibc wrapper; we should instead check whether the returned value is in the valid range of a negative errno value

**Comment by theopolis** [comment]
If I suspect the process `CTRL+Z` then it will "spam" `Lost BPF events counter: 1`. I wonder if it makes sense to add some check to see if the code is continually hitting this branch and avoid logging the same message? That will mean more state tracking and complexity unfortunately.

**Review by theopolis — APPROVED** [review]
Yay, please see my last comment about suspending but we can also address that in a follow up PR. Finally, I think the process might be taking too long to quit `CTRL+D`. Do you think we can make changes and have it exit faster? (also consider this for a follow up)
